### PR TITLE
Fix diff validation and apply

### LINKED ISSUE #16928: Getting many errors when applying schema snapshot from CDB to Postgres
### Describe the Bug

I generated a schema from a directus instance running on CDB using the CLI, and I am trying to apply it on a postgres directus instance. So I am facing multiple errors and so far haven't got it working yet. Errors I got so far:
```
invalid input syntax for type numeric: "NULL"
invalid input syntax for type integer: "nextval('public."CompanySize_id_seq"'::REGCLASS)"
```
First one was easy to fix ad I provided a default `0` instead of the empty null value, although Null is what I need. The second error is harder to resolve as it involves id change, and I don't want to play with that.

Though I would share here before changing my schema anymore.

### To Reproduce

Create some collections and items on CDB and use decimal types with null default and auto increment ids. 
Generate a schema and try to apply it to a Postrgres instance.

### Errors Shown

Multiple:
```
invalid input syntax for type numeric: "NULL"
invalid input syntax for type integer: "nextval('public."CompanySize_id_seq"'::REGCLASS)"
```

### What version of Directus are you using?

9.22.1

### What version of Node.js are you using?

18.2.0

### What database are you using?

CDB v22.1.12, PG15

### What browser are you using?

Chrome

### How are you deploying Directus?

Directus locally, CDB serverless, PG neon

### LINKED ISSUE #17698: Error staging a schema: adding icon and color
### Checklist

- [X] I'm using the [latest version of Directus](https://github.com/directus/directus/releases)
- [X] There's [no other issue](https://github.com/directus/directus/issues?q=is%3Aissue) that already describes the problem.
- [X] I've completed all [Troubleshooting Steps](https://docs.directus.io/getting-started/support/#troubleshooting-steps).

### Describe the Bug

I want to transfer the schema of a Directus instance to another Directus instance.
For this I use the steps as described here: https://docs.directus.io/reference/system/schema.html

We're using Postgres 15.1 and Directus 9.23.1

----

I've changed color and icon of one collection in the source instance. Destination instance doesn't have any color or icon defined for this collection.

Here is the payload after generating the diff (`/schema/diff`) for `/schema/apply`

```
{
    "hash": "06f905c54e2700d9e2167275cf5ce62f3d90566a",
    "diff": {
        "collections": [
            {
                "collection": "test2",
                "diff": [
                    {
                        "kind": "E",
                        "path": [
                            "meta",
                            "color"
                        ],
                        "lhs": null,
                        "rhs": "#FB0404"
                    },
                    {
                        "kind": "E",
                        "path": [
                            "meta",
                            "icon"
                        ],
                        "lhs": null,
                        "rhs": "event"
                    }
                ]
            }
        ],
        "fields": [],
        "relations": []
    }
}
```

Error-Response
```
{
    "errors": [
        {
            "message": "\"diff.collections[0].diff[0].lhs\" must be of type object",
            "extensions": {
                "code": "INVALID_PAYLOAD"
            }
        }
    ]
}
```

In my point of view it shouldn't raise an 'INVALID_PAYLOAD' because there is no color or icon defined in the destination and therefore `lhs` should be null.

### To Reproduce

- Start two instance of directus
- Change color in one of them
- Transfer schema from that instance to the other  as described her https://docs.directus.io/reference/system/schema.html

### Hosting Strategy

Self-Hosted (Docker Image)

### LINKED ISSUE #17820: Failing to apply snapshot schema when altering field sort order
### Checklist

- [X] I'm using the [latest version of Directus](https://github.com/directus/directus/releases)
- [X] There's [no other issue](https://github.com/directus/directus/issues?q=is%3Aissue) that already describes the problem.
- [X] I've completed all [Troubleshooting Steps](https://docs.directus.io/getting-started/support/#troubleshooting-steps).

### Describe the Bug

Getting this exception (`articles` is a data model):

```
error: alter table "articles" alter column "id" drop not null - column "id" is in a primary key
    at Parser.parseErrorMessage (/.../node_modules/pg-protocol/dist/parser.js:287:98)
    at Parser.handlePacket (/.../node_modules/pg-protocol/dist/parser.js:126:29)
    at Parser.parse (/.../node_modules/pg-protocol/dist/parser.js:39:38)
    at Socket.<anonymous> (/.../node_modules/pg-protocol/dist/index.js:11:42)
    at Socket.emit (node:events:513:28)
    at addChunk (node:internal/streams/readable:324:12)
    at readableAddChunk (node:internal/streams/readable:297:9)
    at Readable.push (node:internal/streams/readable:234:10)
    at TCP.onStreamRead (node:internal/stream_base_commons:190:23)
```

It happens after _only_ altering the visual order of a field, i.e. when using drag-and-drop and then updating the schema snapshot file to be shared with other developers. I'd expect that just altering `sort` would not cause a database schema change since it's just a value in the `directus_fields` table.

I'm not 100% certain no other issues also encounter this, but similar stack traces refer to other possible causes.

### To Reproduce

I've setup a repository to reproduce the issue. You can find it here https://github.com/smonn/repro-directus-snapshot-issue

### Hosting Strategy

Self-Hosted (Custom)

### LINKED ISSUE #17832: Foreign keys are set to null when applying schema changes in SQLite
### Checklist

- [X] I'm using the [latest version of Directus](https://github.com/directus/directus/releases)
- [X] There's [no other issue](https://github.com/directus/directus/issues?q=is%3Aissue) that already describes the problem.
- [X] I've completed all [Troubleshooting Steps](https://docs.directus.io/getting-started/support/#troubleshooting-steps).

### Describe the Bug

When using Schema.yaml files to apply data structure changes that remove a not nullable relation, we lose relations to other collections. 
e.g. 

**Collection page**
- image* (directus_file)
- buttons (many-many)

**Collection button**
- name*

When we remove the image by applying a schema change, it will null the relations to buttons in the relation table page_button. It also happens if we do not delete the image field, but allow it to be nullable. 

We use the Directus version 9.23.1 and SQLite3.

Thanks in advance! Please let me know if you need more details.

### To Reproduce

https://gitlab.com/blaemmle/directus-sqlite3-reference-bug

I've tried to create a small Repo with some steps to reproduce the problem. 

### Hosting Strategy

Self-Hosted (Docker Image)

### PR DESCRIPTION
**Pull request by azrikahar**
## Description

Although there are three distinct issues being resolved here, they are co-dependent when applying diffs in general so this PR includes the fixes for all three.

> **Warning**
> - [x] Revert testing for all vendors before merging

### Diff validation

Fixes #17698, fixes ENG-787

Constraining `lhs` and `rhs` as objects are an oversight, as there are many instances that it could be number, string etc. Also updated the `{is: , then: , otherwise: }` Joi rules to more accurately follow [`deep-diff`'s documented logic](https://www.npmjs.com/package/deep-diff#differences):

> `lhs` - the value on the left-hand-side of the comparison (undefined if kind === 'N')
> `rhs` - the value on the right-hand-side of the comparison (undefined if kind === 'D') 

with a minor adjustment to also include `DiffKind.ARRAY` as it doesn't have lhs/rhs.

### Diff apply for field updates

Fixes #17820, fixes ENG-827

Before applying diff, it takes the current snapshot (`currentSnapshot`) and compares it with the current diff (`payload.diff`):

https://github.com/directus/directus/blob/f900e59dcbf687fad9565354eec0f7f47e2f779b/api/src/services/schema.ts#L33-L38

however when retrieving the snapshot, fields are sanitized to remove db vendor-specific properties:

https://github.com/directus/directus/blob/f900e59dcbf687fad9565354eec0f7f47e2f779b/api/src/utils/get-snapshot.ts#L39

but when apply diff updates a field by calling `fieldsService.updateField()`, the changes will be compared with _unsanitized_ field schema (`existingColumn`) as it grabs the value directly from schema inspector:

https://github.com/directus/directus/blob/f900e59dcbf687fad9565354eec0f7f47e2f779b/api/src/services/fields.ts#L396-L407 

This PR adds an additional `sanitizeColumn` util function to prevent faulty false positives in the `isEqual` comparison.

### Diff apply for SQLite

Fixes #17832, fixes ENG-825

Due to applyDiff wrapping everything within a transaction:

https://github.com/directus/directus/blob/f1b48cb81db03986ed19f9d5c198e56981a7aaa8/api/src/utils/apply-diff.ts#L43

SQLite's disabling of foreign keys became no-op:

- https://github.com/directus/directus/blob/f1b48cb81db03986ed19f9d5c198e56981a7aaa8/api/src/database/helpers/schema/dialects/sqlite.ts#L8

- https://github.com/directus/directus/blob/f1b48cb81db03986ed19f9d5c198e56981a7aaa8/api/src/database/helpers/schema/dialects/sqlite.ts#L15

This PR wraps the transaction with the necessary pre/post column changes. This fix is based on #14512 and #15403.

### Diff apply for CockroachDB

Fixes #16928, fixes ENG-352

Currently crdb's `parseDefaultValue` is not accounting for `'NULL'` default value:

https://github.com/directus/directus/blob/e56a6715be8f567d183fe18d3dc56345a9174c84/packages/schema/src/dialects/cockroachdb.ts#L44-L51

whereas Postgres' does with `if (value?.trim().toLowerCase() === 'null')`:

https://github.com/directus/directus/blob/e56a6715be8f567d183fe18d3dc56345a9174c84/packages/schema/src/dialects/postgres.ts#L37-L46

This PR adds the same logic, and resolves the following scenario: When we are only changing a field's translation, the diff used to also affect the field's `schema.default_value` because of null being interpreted as string `'NULL'`, thus creating an unnecessary diff (and wrong diff for that matter).

#### Before

```json
{
	"hash": "23e9ca281509a3978e51c5e1559d9abff61e0445",
	"diff": {
		"collections": [],
		"fields": [
			{
				"collection": "test",
				"field": "metric",
				"diff": [
					{
						"kind": "E",
						"path": ["meta", "translations"],
						"lhs": [
							{
								"language": "en-US",
								"translation": "Metric"
							}
						],
						"rhs": null
					},
					{
						"kind": "E",
						"path": ["schema", "default_value"],
						"lhs": "NULL",
						"rhs": null
					}
				]
			}
		],
		"relations": []
	}
}
```

#### After

```json
{
	"hash": "bc13d501347c17190f69335c189885f1ee167d65",
	"diff": {
		"collections": [],
		"fields": [
			{
				"collection": "test",
				"field": "metric",
				"diff": [
					{
						"kind": "E",
						"path": ["meta", "translations"],
						"lhs": [
							{
								"language": "en-US",
								"translation": "Metric"
							}
						],
						"rhs": null
					}
				]
			}
		],
		"relations": []
	}
}
```

## Type of Change

- [x] Bugfix
- [ ] Improvement
- [ ] New Feature
- [ ] Refactor / codestyle updates
- [ ] Other, please describe:

## Requirements Checklist

- [x] New / updated tests are included
- [x] All tests are passing locally
- [x] Performed a self-review of the submitted code

If adding a new feature:

- [ ] Documentation was added/updated. PR:

### CODE REVIEW

**Comment by azrikahar** [comment]
@licitdev mind taking a look to add blackbox tests for this? 🙏 I believe something similar to https://github.com/directus/directus/pull/15403/files#diff-a312a43e625baea3b14a5d3ce2428bfb518292c43647539b64be83627328c29b but for schema applies, should account for SQLite. Similarly, schema apply with change fields sort order including primary keys for the other 2 scenarios if I'm not mistaken.

**Review by rijkvanzanten — APPROVED** [review]

**Review by paescuj — APPROVED** [review]
Awesome work 🚀 Basically ready for merging, waiting for blackbox tests which would be indeed very valuable here 👍

File: api/src/utils/validate-diff.ts [hunk 1]
```diff
@@ -7,11 +7,15 @@ const deepDiffSchema = Joi.object({
 		.valid(...Object.values(DiffKind))
 		.required(),
 	path: Joi.array().items(Joi.string()),
```
Comment (licitdev) [inline api/src/utils/validate-diff.ts:9]:
```suggestion
	path: Joi.array().items(Joi.alternatives().try(Joi.string(), Joi.number())),
```

The path may contain the integer index when an array value is edited, which currently errors out with `"diff.fields[0].diff[1].path[2]" must be a string`.

### Partial Diff
```js
fields: [
  {
    collection: "test",
    field: "name",
    diff: [
      {
        kind: "A",
        path: ["meta", "translations"],
        index: 2,
        item: {
          kind: "D",
          lhs: { language: "nl-NL", translation: "naam" },
        },
      },
      {
        kind: "E",
        path: ["meta", "translations", 1, "language"],
        lhs: "es-ES",
        rhs: "nl-NL",
      },
      {
        kind: "E",
        path: ["meta", "translations", 1, "translation"],
        lhs: "nombre",
        rhs: "naam",
      },
    ],
  },
];
```

Comment (paescuj) [reply api/src/utils/validate-diff.ts]: Wow, good catch! Thanks ❤️

**Pull request merged**